### PR TITLE
Bumped mypy to 1.10.0 and ruff to 0.4.4

### PR DESCRIPTION
**Pull request by superbobry**

### CODE REVIEW

File: jax/_src/interpreters/pxla.py [hunk 1]
```diff
@@ -2000,7 +2000,7 @@ def are_all_shardings_default_mem_kind(da_object, shardings):
       return False
   return True
 
-memory_kind_propagate_rule = {}  # type: ignore
+memory_kind_propagate_rule: dict[Any, Any] = {}
```
Comment (yashk2810) [inline jax/_src/interpreters/pxla.py:2003]: Why not type: ignore?

Comment (superbobry) [reply jax/_src/interpreters/pxla.py:2003]: `type: ignore` tells the type checker to treat this as `Any`, but we know it's at least a `dict`.

Comment (yashk2810) [reply jax/_src/interpreters/pxla.py:2003]: But why did that fail on this PR where you are just bumping the version?

Comment (superbobry) [reply jax/_src/interpreters/pxla.py:2003]: Newer mypy emits an error when `memory_kind_propagate_rule` is modified.

**Comment by jakevdp** [comment]
It looks like the new versions are sensitive to some lint and typecheck errors that will need to be addressed. Thanks!

**Comment by superbobry** [comment]
Yep, I'm working on updating existing code. 

I also really like that ruff now flags `sum([x for x in xs])`!

File: jax/_src/interpreters/pxla.py (hunk 1 above)
Comment (yashk2810) [reply jax/_src/interpreters/pxla.py:2003]: Hmm.. interesting. OOC, what's the error?

**Comment by superbobry** [comment]
I think this is good to go.

**Review by jakevdp — APPROVED** [review]
Looks good!

**Pull request merged**